File: adventjs/2025/23/solution.ts
function minStepsToDeliver(map: string[][]): number {
  const rows = map.length, cols = map[0].length;
  let startR = -1, startC = -1;
  const goals: [number, number][] = [];
  
  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < cols; c++) {
      if (map[r][c] === 'S') {
        startR = r;
        startC = c;
      } else if (map[r][c] === 'G') {
        goals.push([r, c]);
      }
    }
  }
  
  if (goals.length === 0) return 0;
  
  const isValid = (r: number, c: number): boolean => 
    r >= 0 && r < rows && c >= 0 && c < cols && map[r][c] !== '#';
  
  const bfs = (sr: number, sc: number, gr: number, gc: number): number => {
    const queue: [number, number, number][] = [[sr, sc, 0]];
    const visited = new Set([`${sr},${sc}`]);
    const dirs: [number, number][] = [[-1, 0], [1, 0], [0, -1], [0, 1]];
    
    while (queue.length) {
      const [r, c, steps] = queue.shift()!;
      if (r === gr && c === gc) return steps;
      
      for (const [dr, dc] of dirs) {
        const nr = r + dr, nc = c + dc;
        const key = `${nr},${nc}`;
        
        if (isValid(nr, nc) && !visited.has(key)) {
          visited.add(key);
          queue.push([nr, nc, steps + 1]);
        }
      }
    }
    return -1;
  };
  
  let total = 0;
  for (const [gr, gc] of goals) {
    const dist = bfs(startR, startC, gr, gc);
    if (dist === -1) return -1;
    total += dist;
  }
  
  return total;
}

File: adventjs/2025/23/solution.py
def min_steps_to_deliver(map: list[list[str]]) -> int:
  rows, cols = len(map), len(map[0])
  start_r, start_c = -1, -1
  goals = []
  
  for r in range(rows):
    for c in range(cols):
      if map[r][c] == 'S':
        start_r, start_c = r, c
      elif map[r][c] == 'G':
        goals.append((r, c))
  
  if not goals:
    return 0
  
  def bfs(sr, sc, gr, gc):
    queue = [(sr, sc, 0)]
    visited = {(sr, sc)}
    dirs = [(-1, 0), (1, 0), (0, -1), (0, 1)]
    
    while queue:
      r, c, steps = queue.pop(0)
      
      if r == gr and c == gc:
        return steps
      
      for dr, dc in dirs:
        nr, nc = r + dr, c + dc
        if 0 <= nr < rows and 0 <= nc < cols and map[nr][nc] != '#' and (nr, nc) not in visited:
          visited.add((nr, nc))
          queue.append((nr, nc, steps + 1))
    return -1
  
  total = 0
  for gr, gc in goals:
    dist = bfs(start_r, start_c, gr, gc)
    if dist == -1:
      return -1
    total += dist
  
  return total
